spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/SparkPositionDeltaWriteBuilder.java

```diff
@@ -0,0 +1,123 @@
+import org.apache.spark.sql.connector.read.Scan;
+import org.apache.spark.sql.types.StructType;
+
+class SparkPositionDeltaWriteBuilder implements DeltaWriteBuilder {
+
+  private final SparkSession spark;
+  private final Table table;
+  private final Command command;
+  private final SparkBatchQueryScan scan;
+  private final IsolationLevel isolationLevel;
+  private final SparkWriteConf writeConf;
+  private final ExtendedLogicalWriteInfo info;
+  private final boolean handleTimestampWithoutZone;
+  private final boolean checkNullability;
+  private final boolean checkOrdering;
+
+  SparkPositionDeltaWriteBuilder(SparkSession spark, Table table, Command command, Scan scan,
+                                 IsolationLevel isolationLevel, ExtendedLogicalWriteInfo info) {
+    this.spark = spark;
+    this.table = table;
+    this.command = command;
+    this.scan = (SparkBatchQueryScan) scan;
+    this.isolationLevel = isolationLevel;
+    this.writeConf = new SparkWriteConf(spark, table, info.options());
+    this.info = info;
+    this.handleTimestampWithoutZone = writeConf.handleTimestampWithoutZone();
+    this.checkNullability = writeConf.checkNullability();
+    this.checkOrdering = writeConf.checkOrdering();
+  }
+
+  @Override
+  public DeltaWrite build() {
+    Preconditions.checkArgument(handleTimestampWithoutZone || !SparkUtil.hasTimestampWithoutZone(table.schema()),
+        SparkUtil.TIMESTAMP_WITHOUT_TIMEZONE_ERROR);
+
+    Schema dataSchema = dataSchema();
+    if (dataSchema != null) {
+      TypeUtil.validateWriteSchema(table.schema(), dataSchema, checkNullability, checkOrdering);
+    }
+
+    Schema expectedRowIdSchema = expectedRowIdSchema();
+    Schema rowIdSchema = SparkSchemaUtil.convert(expectedRowIdSchema, info.rowIdSchema());
+    TypeUtil.validateSchema("row ID", expectedRowIdSchema, rowIdSchema, checkNullability, checkOrdering);
+
+    Schema expectedMetadataSchema = expectedMetadataSchema();
+    Schema metadataSchema = SparkSchemaUtil.convert(expectedMetadataSchema, info.metadataSchema());
+    TypeUtil.validateSchema("metadata", expectedMetadataSchema, metadataSchema, checkNullability, checkOrdering);
+
+    SparkUtil.validatePartitionTransforms(table.spec());
+
+    Distribution distribution = SparkDistributionAndOrderingUtil.buildPositionDeltaDistribution(
+        table, command, distributionMode());
+    SortOrder[] ordering = SparkDistributionAndOrderingUtil.buildPositionDeltaOrdering(
+        table, command, distribution);
+
+    return new SparkPositionDeltaWrite(
+        spark, table, command, scan, isolationLevel, writeConf,
+        info, dataSchema, distribution, ordering);
+  }
+
+  private Schema dataSchema() {
+    StructType dataSparkType = info.schema();
+    return dataSparkType != null ? SparkSchemaUtil.convert(table.schema(), dataSparkType) : null;
+  }
+
+  private Schema expectedRowIdSchema() {
+    return new Schema(MetadataColumns.FILE_PATH, MetadataColumns.ROW_POSITION);
```

Comment:
Looks like this could be a static final field.

Reply:
Done.